File: src/main/java/im/joker/helper/EventSyncQueueManager.java
package im.joker.helper;

import com.google.common.collect.Maps;
import im.joker.event.room.AbstractRoomEvent;
import im.joker.util.GsonUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.ReactiveStringRedisTemplate;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.util.function.Tuple2;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static im.joker.constants.ImRedisKeys.ACTIVE_ROOM_LATEST_EVENTS;

@Component
public class EventSyncQueueManager {

    @Autowired
    private ReactiveStringRedisTemplate redisTemplate;

    private final int limit = 100;
    @Autowired
    private RoomSubscribeManager roomSubscribeManager;
    @Autowired
    private RequestProcessor requestProcessor;

    /**
     * 从对应的设备中到自己关心的队列里,拿出对应限制条数的消息.是有序的
     * key是房间, value是此房间的事件集
     *
     * @param deviceId
     * @param limitOfRoom
     * @return
     */
    public Mono<Map<String, List<AbstractRoomEvent>>> takeRelatedEvents(String deviceId, int limitOfRoom) {
        Flux<String> careRoomIds = roomSubscribeManager.retrieveRooms(deviceId);
        return careRoomIds.flatMap(roomId -> {
            return Mono.just(roomId)
                    .zipWith(redisTemplate.opsForList().range(String.format(ACTIVE_ROOM_LATEST_EVENTS, roomId), 0, limitOfRoom)
                            .map(s -> requestProcessor.toBean(s, AbstractRoomEvent.class))
                            .collectList());
        }).collectSortedList()
                .map(tuple2s -> {
                    Map<String, List<AbstractRoomEvent>> retMap = Maps.newHashMap();
                    tuple2s.forEach(tuple2 -> retMap.put(tuple2.getT1(), tuple2.getT2()));
                    return retMap;
                });
    }


    /**
     * 将消息存到sync时的消息队列,并且每个房间只维持最新的100条数据
     *
     * @param roomEvent
     * @return
     */
    public Mono<Void> addEventToQueue(AbstractRoomEvent roomEvent) {

        return redisTemplate.opsForList().rightPush(String.format(ACTIVE_ROOM_LATEST_EVENTS, roomEvent.getRoomId()), GsonUtils.get().toJson(roomEvent))
                .flatMap(e -> redisTemplate.opsForList().size(String.format(ACTIVE_ROOM_LATEST_EVENTS, roomEvent.getRoomId())))
                .flatMap(size -> {
                    if (size > limit) {
                        return redisTemplate.opsForList().trim(String.format(ACTIVE_ROOM_LATEST_EVENTS, roomEvent.getRoomId()), size - limit, -1);
                    }
                    return Mono.empty();
                })
                .then();

    }


}

File: src/main/kotlin/im/joker/helper/EventSyncQueueManager.kt
package im.joker.helper

import im.joker.constants.ImConstants.Companion.ACTIVE_ROOM_LATEST_EVENTS
import im.joker.event.room.AbstractRoomEvent
import im.joker.event.room.AbstractRoomStateEvent
import im.joker.event.room.other.FullReadMarkerEvent
import im.joker.event.room.other.ReceiptEvent
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.reactive.awaitFirstOrNull
import kotlinx.coroutines.reactive.awaitSingle
import kotlinx.coroutines.reactive.awaitSingleOrNull
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.redis.core.ReactiveStringRedisTemplate
import org.springframework.stereotype.Component

@Component
class EventSyncQueueManager {

    private val log: Logger = LoggerFactory.getLogger(EventSyncQueueManager::class.java)

    @Autowired
    private lateinit var redisTemplate: ReactiveStringRedisTemplate

    // 每个房间的事件队列为n条消息
    private val limit = 100

    @Autowired
    private lateinit var roomSubscribeManager: RoomSubscribeManager

    @Autowired
    private lateinit var requestProcessor: RequestProcessor


    /**
     * 将事件放入redis对应的room队列
     */
    suspend fun addToEventQueue(ev: AbstractRoomEvent) {
//        log.debug("将ev放入消息队列:{}", requestProcessor.toJson(ev))
        redisTemplate.opsForList().rightPush(ACTIVE_ROOM_LATEST_EVENTS.format(ev.roomId), requestProcessor.toJson(ev)).awaitSingleOrNull()
        val queueSize = redisTemplate.opsForList().size(ACTIVE_ROOM_LATEST_EVENTS.format(ev.roomId)).awaitSingle()
        // 当超过limit条事件的时候,修整该房间队列,删除最老的消息,以维持队列保证limit条
        if (limit > queueSize) {
            redisTemplate.opsForList().trim(ACTIVE_ROOM_LATEST_EVENTS.format(ev.roomId), queueSize - limit, -1).awaitSingleOrNull()
        }
    }

    /**
     * 从设备相关的房间中拿取limitOfRoom条数据,其返回值key是roomId,value是redis队列中提取的
     * 过滤其本人说的话
     */
    suspend fun takeRelatedEvent(deviceId: String, sender: String, gteStreamId: Long, lteStreamId: Long): Map<String, List<AbstractRoomEvent>> = coroutineScope {
        val roomIds = roomSubscribeManager.searchJoinRoomIds(deviceId)
        val list = roomIds.map {
            async {
                redisTemplate.opsForList().range(ACTIVE_ROOM_LATEST_EVENTS.format(it), 0, -1).collectList().awaitSingleOrNull()
            }
        }
        val eventMap = HashMap<String, List<AbstractRoomEvent>>()
        list.awaitAll().forEach { it1 ->
            val roomEvents = it1.map { requestProcessor.toBean(it, AbstractRoomEvent::class.java) }
                    .filter {
                        val range = it.streamId in gteStreamId..lteStreamId
                        if (it is FullReadMarkerEvent) {
                            return@filter range && it.sender == sender
                        }
                        range
                    }
            if (roomEvents.isNotEmpty()) eventMap[roomEvents[0].roomId] = roomEvents
        }
        return@coroutineScope eventMap
    }

}

